### ISSUE TITLE
Using a dialog link breaks the frontend theme display

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#5250]
**Description of the bug**

Sort of related to #4863, but slightly different.

When opening a link to a path that has the admin theme applied, the frontend theme display gets messed. The header looks "sevenish" and so does the menu.

**Steps To Reproduce**

On a vanilla install, where the admin theme is also used for node forms (default).

1. Create a new post, set its format to "Raw HTML" (to insert markup)
2. Make sure that comments are enabled, so the library is loaded (otherwise this simple example won't use ajax)
3. Insert a dialog link: `<a href="/node/1/edit" class="use-ajax" data-dialog="true">dialog link to edit node 1</a>`
4. Save the node
5. Click on that link
6. Close the dialog with the close button on top
7. See the frontend theme messed

**Screenshot after closing**

![broken-theme](https://user-images.githubusercontent.com/842066/134293367-a62d5612-1d29-40a7-a3b5-b087ff2f5560.png)

**Expected behavior**

The page display stays untouched. IMO the dialog should _not_ load another theme's CSS, but always stick with the active one on the page the dialog link is.

**Additional information**

- Backdrop CMS version: latest stable

A workaround is to uncheck "Use the administration theme when editing or creating content".

**Comment by indigoxela** [comment]
Another (smaller) glitch in this special case - node form - is that the cancel link is in the dialog content, not in the dialog button pane.

And even smaller, but still a glitch - the dialog buttons are "jumpy" in Basis, but that's not relevant here.

**Comment by laryn** [comment]
I guess maybe because `/node/%/edit` is considered an admin path, it pulls in admin theme CSS?

See also: https://github.com/backdrop/backdrop-issues/issues/4863

**Comment by laryn** [comment]
Does the page underneath the modal jump as well? https://github.com/backdrop/backdrop-issues/issues/4342

**Comment by indigoxela** [comment]
> I guess maybe because /node/%/edit is considered an admin path

Yes, :wink:, that's what I consider the bug here. Where does it ever make sense, that a dialog *on the same page* loads css from a different theme - theme not active on the page the dialog is shown?

> Does the page underneath the modal jump as well?

Yes, of course. That's an annoying, well-known, but unrelated bug.

**Comment by laryn** [comment]
> Yes, of course. That's an annoying, well-known, but unrelated bug.

Just making sure you were properly annoyed. 😉

**Comment by indigoxela** [comment]
Played around a bit - Ajax isn't my dojo.

Interesting comment for [function ajax_base_page_theme()](https://github.com/backdrop/backdrop/blob/1.x/core/includes/ajax.inc#L422)

... which is exactly why I think switching the theme when loading a dialog is a bug.

The switch happens in function function ajax_render(), where the correct css files ship with `$_POST['ajax_page_state']`, but all of a sudden the css files from seven (admin theme) get loaded with backdrop_get_css(). No clue why that happens.

Function ajax_base_page_theme() doesn't run at all when opening a dialog. Maybe that's the problem (or maybe not).

I guess, I have to wait until someone with better Ajax skills gets as annoyed as me by that bug. :stuck_out_tongue_winking_eye:

**Comment by klonos** [comment]
This affects some contrib that implement `hook_admin_paths()` and/or `hook_admin_paths_alter()`. I recently came across this with https://github.com/backdrop-contrib/diff which has this "Use administration theme" option, which is on by default:

![image](https://user-images.githubusercontent.com/2423362/135373518-b01f00c5-16d2-4ee2-ac4d-7cc0524e833c.png)

That option currently is broken (see https://github.com/backdrop-contrib/diff/issues/17), but once you apply the fix, this is what you get when you are viewing a node revision diff:

![image](https://user-images.githubusercontent.com/2423362/135373662-801c8863-a235-4145-99cc-e9a285331e77.png)

^^ that's supposed to be Seven, but it seems to have elements from Basis 🤷🏼 

Not core-related, but that now gives us another use case to consider and test with 😉

**Comment by indigoxela** [comment]
Just ran into that problem again. I really hope, someone can fix it. I wasn't able to collect more info than we already have. Just have another workaround. :disappointed:

**Comment by argiepiano** [comment]
Hey @indigoxela. Since we recently worked on a dialog/ajax/css related bug with the image browser, I thought I'd take a look at this. But I can't reproduce the problem following your steps in the OP (the edit page loads normally, rather than through a dialog). I'm thinking that there may be a missing step, or perhaps you are using a version of Basis that loads the `backdrop.ajax` system library when viewing a node (as explained [here](https://docs.backdropcms.org/change-records/dialog-api). Is that the case?

**Comment by indigoxela** [comment]
It's still reproducible, but I have to update the description. One step is missing.

**Comment by argiepiano** [comment]
I finally found the reason for this collision of css classes between front and admin themes when ajax is used to load a page that has a different theme from the one being viewed. This is not limited to the issue mentioned in the OP, but to all such cases.

The way to avoid this is by defining a `'theme callback' => 'ajax_base_page_theme'` in the menu item that's causing the problem. If you add this to the `node/%node/edit` item definition, the problem goes away. 

There is a full explanation of this callback in the doc block for `ajax_base_page_theme()`. This callback will not affect the "regular" use of the item - it only prevents the loading of the wrong css when using it as a modal dialog.  It's used in many menu items provided by `layout`.

The question is: do we want to add this to all menu items that could possibly become used in a modal? While harmless, I'm leaning toward a "no" - the possibilities are just too many. It's probably best to let the developer deal with his with a hook_menu_alter. 

Thoughts?

**Comment by indigoxela** [comment]
> I'm leaning toward a "no" - the possibilities are just too many

@argiepiano that seems reasonable to me. :+1: Given the lack of additional feedback here, the problem is rather an edge case, anyway.
